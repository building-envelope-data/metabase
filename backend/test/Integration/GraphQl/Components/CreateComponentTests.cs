using System;
using System.IO;
using System.Threading.Tasks;
using FluentAssertions;
using Metabase.GraphQl.Components;
using Metabase.Tests.Integration.GraphQl.Institutions;
using Snapshooter;
using Snapshooter.NUnit;
using NUnit.Framework;

namespace Metabase.Tests.Integration.GraphQl.Components
{
    [TestFixture]
    public sealed class CreateComponentTests
      : ComponentIntegrationTests
    {
        [Test]
        public async Task AnonymousUser_IsAuthenticationError()
        {
            // Act
            var response =
                await UnsuccessfullyQueryGraphQlContentAsString(
                    File.ReadAllText("Integration/GraphQl/Components/CreateComponent.graphql"),
                    variables: MinimalComponentInput
                ).ConfigureAwait(false);
            // Assert
            Snapshot.Match(response);
        }

        [Test]
        public async Task AnonymousUser_CannotCreateComponent()
        {
            // Act
            await UnsuccessfullyQueryGraphQlContentAsString(
                File.ReadAllText("Integration/GraphQl/Components/CreateComponent.graphql"),
                variables: MinimalComponentInput
                ).ConfigureAwait(false);
            var response = await GetComponents().ConfigureAwait(false);
            // Assert
            Snapshot.Match(response);
        }

        [TestCaseSource(nameof(EnumerateComponentInputs))]
        [Theory]
        public async Task LoggedInUser_IsSuccess(
            string key,
            CreateComponentInput input
        )
        {
            SnapshotFullName testName = SnapshotFullNameHelper(typeof(CreateComponentTests), key);
 
            // Arrange
            var userId = await RegisterAndConfirmAndLoginUser().ConfigureAwait(false);
            var institutionId = await InstitutionIntegrationTests.CreateAndVerifyInstitutionReturningUuid(
                HttpClient,
                InstitutionIntegrationTests.PendingInstitutionInput with
                {
                    OwnerIds = new[] { userId }
                }
                ).ConfigureAwait(false);
            // Act
            var response = await CreateComponent(
                input with
                {
                    ManufacturerId = institutionId
                }
                ).ConfigureAwait(false);
            // Assert
            Snapshot.Match(
                response,
                testName,
                matchOptions => matchOptions
                .Assert(fieldOptions =>
                 fieldOptions.Field<string>("data.createComponent.component.id").Should().NotBeNullOrWhiteSpace()
                 )
                 .Assert(fieldOptions =>
                 fieldOptions.Field<Guid>("data.createComponent.component.uuid").Should().NotBe(Guid.Empty)
                 )
                );
        }

        [TestCaseSource(nameof(EnumerateComponentInputs))]
        [Theory]
        public async Task LoggedInUser_CreatesComponent(
            string key,
            CreateComponentInput input
        )
        {
            SnapshotFullName testName = SnapshotFullNameHelper(typeof(CreateComponentTests), key);
 

            // Arrange
            var userId = await RegisterAndConfirmAndLoginUser().ConfigureAwait(false);
            var institutionId = await InstitutionIntegrationTests.CreateAndVerifyInstitutionReturningUuid(
                HttpClient,
                InstitutionIntegrationTests.PendingInstitutionInput with
                {
                    OwnerIds = new[] { userId }
                }
                ).ConfigureAwait(false);
            // Act
            var (componentId, componentUuid) = await CreateComponentReturningIdAndUuid(
                input with
                {
                    ManufacturerId = institutionId
                }
                ).ConfigureAwait(false);
            var response = await GetComponents().ConfigureAwait(false);
            // Assert
            Snapshot.Match(
                response,
                testName,
                matchOptions => matchOptions
                .Assert(fieldOptions =>
                 fieldOptions.Field<string>("data.components.edges[*].node.id").Should().Be(componentId)
                 )
                .Assert(fieldOptions =>
                 fieldOptions.Field<Guid>("data.components.edges[*].node.uuid").Should().Be(componentUuid)
                 )
                );
        }
    }
}
