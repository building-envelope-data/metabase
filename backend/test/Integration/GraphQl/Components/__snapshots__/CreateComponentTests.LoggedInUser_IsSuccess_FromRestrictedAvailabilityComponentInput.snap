{"data":{"createComponent":{"component":{"id":"Q29tcG9uZW50Cmc2YWIxNDNhM2RkMjI0MDg1ODE2NTM4NjAxZjZmYjc1Yg==","uuid":"6ab143a3-dd22-4085-8165-38601f6fb75b","name":"Component D","abbreviation":"C!D","description":"Whatever component!","availability":{"from":"2019-04-03T00:00:00.000Z","to":null},"categories":["LAYER","MATERIAL","UNIT"]},"errors":null}}}
